File: mixin-base/src/main/python/mixin/model/basic_elist.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.#
from typing import TypeVar, List
from mixin.model.elist import EList
from mixin.model.eutils import EUtils
from mixin.model.eobject import EObject
from mixin.model.efeature import EFeature
from mixin.model.notification import Notification

E = TypeVar("E")


class BasicEList(EList[E]):
    def __init__(self, owner: EObject, feature: EFeature):
        super().__init__(owner, feature)
        self._elements: List[E] = []

    def isEmpty(self) -> bool:
        return len(self._elements) == 0

    def contains(self, object: any) -> bool:
        return object in self._elements

    def get(self, index: int) -> E:
        if index >= 0 or index < len(self._elements):
            return self._elements[index]
        else:
            return None

    def set(self, index: int, element: E) -> E:
        self._deleteFromContainer(element)
        oldElement: E = self._elements[index]
        self._elements[index] = element
        self._owner.notify(Notification.SET, self._feature, oldElement, element, index, None)
        return element

    def _deleteFromContainer(self, element: E):
        if isinstance(element, EObject):
            eObject: EObject = element
            EUtils.remove(eObject)

    def add(self, element: E):
        self._deleteFromContainer(element)
        self._elements.append(element)
        position: int = len(self._elements) - 1
        self._owner.notify(Notification.ADD, self._feature, None, element, position, None)

    def addAt(self, index: int, element: E):

        self._deleteFromContainer(element)

        if index == len(self._elements):

            # Lakukan penambahan di bagian akhir
            self._elements.append(element)
            self._owner.notify(Notification.ADD, self._feature, None, element, index, None)

        elif index >= 0 and index < len(self._elements):

            # Lakukan penambahan di index tersebut
            self._elements.insert(index, element)
            self._owner.notify(Notification.ADD, self._feature, None, element, index, None)

    def remove(self, element: any):
        index: int = self.indexOf(element)
        if index != -1:
            self.removeAt(index)
        return index

    def removeAt(self, index: int) -> E:
        removed: E = self.get(index)
        if removed is not None:
            del self._elements[index]
            self._owner.notify(Notification.REMOVE, self._feature, removed, None, index, None)
        return removed

    def move(self, element: E, index: int):
        position: int = self.indexOf(element)
        if position != -1:
            removed = self._elements[position]
            del self._elements[position]
            self._elements.insert(index, removed)
            self._owner.notify(Notification.MOVE, self._feature, element, element, index, None)

    def indexOf(self, element: any) -> int:
        if element in self._elements:
            return self._elements.index(element)
        else:
            return -1

    def size(self) -> int:
        return len(self._elements)

    def _doClear(self) -> List[E]:
        elements: List[E] = []
        elements.extend(self._elements)
        self._elements.clear()
        return elements

    def addAll(self, source: List[E]):
        index = len(self._elements)
        for element in source:
            self._deleteFromContainer(element)
        self._elements.extend(source)
        self._owner.notify(Notification.ADD_MANY, self._feature, None, source, index, None)

    def insertRange(self, source: List[E], start: int):
        for element in source:
            self._deleteFromContainer(element)
        self._elements[start:start] = source
        self._owner.notify(Notification.ADD_MANY, self._feature, None, source, start, None)

    def removeRange(self, start: int, end: int):
        removed: List[E] = []
        for i in range(start, end):
            element: E = self._elements[i]
            removed.append(element)
        for remove in removed:
            self._elements.remove(remove)
        self._owner.notify(Notification.REMOVE_MANY, self._feature, removed, None, start, None)

    def clear(self):
        elements: List[E] = self._doClear()
        self._owner.notify(Notification.REMOVE_MANY, self._feature, elements, None, -1, None)

    def repopulate(self, source: List[E]):
        removes: List[E] = self._doClear()
        self._elements.extend(source)
        self._owner.notify(Notification.REPLACE_MANY, self._feature, removes, source, 0, None)

    def __iter__(self):
        return self._elements.__iter__()

    def toArray(self) -> List[any]:
        return self._elements

File: mixin-base/src/main/java/com/andia/mixin/model/BasicEList.java
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. */
package com.andia.mixin.model;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

public class BasicEList<E> extends EList<E> {

	private List<E> elements = new ArrayList<>();

	public BasicEList(EObject owner, EFeature feature) {
		super(owner, feature);
	}

	@Override
	public boolean isEmpty() {
		return elements.isEmpty();
	}

	@Override
	public boolean contains(Object o) {
		return elements.contains(o);
	}

	@Override
	public E get(int index) {
		return this.elements.get(index);
	}

	@Override
	public E set(int index, E element) {
		this.deleteFromContainer(element);
		E oldElement = this.elements.get(index);
		this.elements.set(index, element);
		this.owner.notify(Notification.SET, this.feature, oldElement, element, index, null);
		return element;
	}

	private void deleteFromContainer(E element) {
		if (element instanceof EObject) {
			EObject eObject = (EObject) element;
			EUtils.remove(eObject);
		}
	}

	@Override
	public void add(E element) {
		this.deleteFromContainer(element);
		this.elements.add(element);
		int position = this.elements.size() - 1;
		this.owner.notify(Notification.ADD, this.feature, null, element, position, null);
	}

	@Override
	public void add(int index, E element) {

		this.deleteFromContainer(element);

		if (index == this.elements.size()) {

			// Lakukan penambahan di bagian akhir
			this.elements.add(element);
			this.owner.notify(Notification.ADD, this.feature, null, element, index, null);

		} else if (index >= 0 && index < this.elements.size()) {

			// Lakukan penambahan di index tersebut
			this.elements.add(index, element);
			this.owner.notify(Notification.ADD, this.feature, null, element, index, null);

		}
	}

	@Override
	public void remove(Object element) {
		int index = this.elements.indexOf(element);
		remove(index);
	}

	@Override
	public E remove(int index) {
		E removed = this.elements.remove(index);
		if (removed != null) {
			this.owner.notify(Notification.REMOVE, this.feature, removed, null, index, null);
		}
		return removed;
	}

	@Override
	public void move(E element, int index) {
		int position = this.elements.indexOf(element);
		if (position != -1) {
			E removed = this.elements.remove(position);
			this.elements.add(index > position ? index - 1 : index, removed);
			this.owner.notify(Notification.MOVE, this.feature, element, element, index, null);
		}
	}

	@Override
	public int indexOf(Object element) {
		return this.elements.indexOf(element);
	}

	@Override
	public int size() {
		return this.elements.size();
	}

	private List<E> doClear() {
		List<E> elements = new ArrayList<E>();
		elements.addAll(this.elements);
		this.elements.clear();
		return elements;
	}

	@Override
	public void addAll(Collection<? extends E> source) {
		doAddAll(source);
		this.owner.notify(Notification.ADD_MANY, this.feature, null, source, -1, null);
	}

	private void doAddAll(Collection<? extends E> source) {
		for (E element : source) {
			this.deleteFromContainer(element);
		}
		this.elements.addAll(source);
	}

	@Override
	public void insertRange(Collection<? extends E> source, int start) {
		for (E element : source) {
			this.deleteFromContainer(element);
		}
		this.elements.addAll(start, source);
		this.owner.notify(Notification.ADD_MANY, this.feature, null, source, start, null);
	}

	@Override
	public void removeRange(int start, int end) {
		List<E> removed = new ArrayList<>();
		for (int i = start; i < end; i++) {
			E e = this.elements.get(i);
			removed.add(e);
		}
		this.elements.removeAll(removed);
		this.owner.notify(Notification.REMOVE_MANY, this.feature, removed, null, start, null);
	}

	@Override
	public void clear() {
		List<E> elements = doClear();
		this.owner.notify(Notification.REMOVE_MANY, this.feature, elements, null, -1, null);
	}

	@Override
	public void repopulate(Collection<? extends E> source) {
		List<E> removes = doClear();
		this.elements.addAll(source);
		this.owner.notify(Notification.REPLACE_MANY, this.feature, removes, source, -1, null);
	}

	@Override
	public Iterator<E> iterator() {
		return this.elements.iterator();
	}

	@Override
	public Object[] toArray() {
		return this.elements.toArray();
	}

}
